core/src/banking_stage/worker.rs

```diff
@@ -0,0 +1,631 @@
+    }
+
+    #[test]
+    fn test_worker_consume_simple() {
+        let (test_frame, worker) = setup_test_frame();
+        let TestFrame {
+            mint_keypair,
+            genesis_config,
+            bank,
+            poh_recorder,
+            consume_sender,
+            consumed_receiver,
+            ..
+        } = &test_frame;
+        let worker_thread = std::thread::spawn(move || worker.run());
+        poh_recorder.write().unwrap().set_bank(bank, false);
+
+        let pubkey1 = Pubkey::new_unique();
+
+        let transactions = sanitize_transactions(vec![system_transaction::transfer(
+            mint_keypair,
+            &pubkey1,
+            1,
+            genesis_config.hash(),
+        )]);
+        let bid = TransactionBatchId::new(0);
+        let id = TransactionId::new(0);
+        let work = ConsumeWork {
+            batch_id: bid,
+            ids: vec![id],
+            transactions,
+            max_age_slots: vec![bank.slot()],
+        };
+        consume_sender.send(work).unwrap();
+        let consumed = consumed_receiver.recv().unwrap();
+        assert_eq!(consumed.work.batch_id, bid);
+        assert_eq!(consumed.work.ids, vec![id]);
+        assert_eq!(consumed.work.max_age_slots, vec![bank.slot()]);
+        assert_eq!(consumed.retryable_indexes, Vec::<usize>::new());
+
+        drop(test_frame);
+        let _ = worker_thread.join().unwrap();
+    }
+
+    #[test]
+    fn test_worker_consume_self_conflicting() {
+        let (test_frame, worker) = setup_test_frame();
+        let TestFrame {
+            mint_keypair,
+            genesis_config,
+            bank,
+            poh_recorder,
+            consume_sender,
+            consumed_receiver,
+            ..
+        } = &test_frame;
+        let worker_thread = std::thread::spawn(move || worker.run());
+        poh_recorder.write().unwrap().set_bank(bank, false);
+
+        let pubkey1 = Pubkey::new_unique();
+        let pubkey2 = Pubkey::new_unique();
+
+        let txs = sanitize_transactions(vec![
+            system_transaction::transfer(mint_keypair, &pubkey1, 2, genesis_config.hash()),
+            system_transaction::transfer(mint_keypair, &pubkey2, 2, genesis_config.hash()),
+        ]);
+
+        let bid = TransactionBatchId::new(0);
+        let id1 = TransactionId::new(1);
+        let id2 = TransactionId::new(0);
+        consume_sender
+            .send(ConsumeWork {
+                batch_id: bid,
+                ids: vec![id1, id2],
+                transactions: txs,
+                max_age_slots: vec![bank.slot(), bank.slot()],
+            })
+            .unwrap();
+
+        let consumed = consumed_receiver.recv().unwrap();
+        assert_eq!(consumed.work.batch_id, bid);
+        assert_eq!(consumed.work.ids, vec![id1, id2]);
+        assert_eq!(consumed.work.max_age_slots, vec![bank.slot(), bank.slot()]);
+        assert_eq!(consumed.retryable_indexes, vec![1]); // id2 is retryable since lock conflict
```

Comment:
This is not a *realistic* case, as the scheduler should not be sending self-conflicting batches. But the worker is able to gracefully handle if the scheduler does.